dynamic/internal/shim/grpcutil/replay.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcutil
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/opentofu/opentofu/internal/tfplugin6"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"google.golang.org/grpc"
+)
+
+// LogReplayProvider is a provider that replays logs from a previous run.
+type LogReplayProvider struct {
+	methodLogs map[string][]grpcLog
+}
+
+var _ = tfplugin6.ProviderClient(LogReplayProvider{})
+
+func getMethodFromFullName(fullName string) string {
+	splitName := strings.Split(fullName, ".")
+	return splitName[len(splitName)-1]
+}
+
+// NewLogReplayProvider creates a new LogReplayProvider from the given logs.
+// It uses the logs to replay the recorded calls to the provider.
+func NewLogReplayProvider(name, version string, logs []byte) LogReplayProvider {
+	var grpcLogs []grpcLog
+	err := json.Unmarshal(logs, &grpcLogs)
+	contract.AssertNoErrorf(err, "failed to unmarshal logs")
+	methodLogs := make(map[string][]grpcLog)
+	for _, log := range grpcLogs {
+		methodName := getMethodFromFullName(log.Name)
+		methodLogs[methodName] = append(methodLogs[methodName], log)
+	}
+	return LogReplayProvider{
+		methodLogs: methodLogs,
+	}
+}
+
+func (p LogReplayProvider) mustPopLog(method string) grpcLog {
+	logs, ok := p.methodLogs[method]
+	contract.Assertf(ok && len(logs) > 0, "no logs for method %s, logs: %s", method, p.methodLogs)
+	log := logs[0]
```

Comment:
The ordering of incoming requests will not be reliable in the general case. This will need a customizable matcher strategy from request to which log, and the default one probably should be a bit more aggressive than just matching on method + position. Not blocking on this but this will 100% bite us and need further elaboration.

Reply:
I haven't hit any issues with this yet and have opted for something simple here and we can extend later.

Reply:
I'd like to push back on this. This is 100% programming in pain for maintainers. Can we at least have a tracking issue with cross-linked TODO in the code?

Reply:
Opened https://github.com/pulumi/pulumi-terraform-bridge/issues/2571